### PR TITLE
Flutter vulnerable module for Ostorlab insecure app

### PR DESCRIPTION
**Pull request by BlueSquare1**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -122,4 +123,7 @@ captures/
 
 # Intellij
 *.iml
-.idea/workspace.xml
\ No newline at end of file
+.idea/workspace.xml
+
+# Flutter module
+!ostorlab_insecure_flutter_app/**
```
Comment (3asm) [inline .gitignore]: Why adding this?

File: app/build.gradle [hunk 2]
```diff
@@ -37,19 +40,36 @@ android {
     }
 }
 
+String storageUrl = System.env.FLUTTER_STORAGE_BASE_URL ?:
+        "https://storage.googleapis.com"
+repositories {
+    maven {
+        url '../ostorlab_insecure_flutter_app/build/host/outputs/repo'
+    }
```
Comment (3asm) [inline app/build.gradle:48]: Why this?

File: ostorlab_insecure_flutter_app/README.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+A new Flutter module project.
+
+## Getting Started
+
+For help getting started with Flutter development, view the online
+[documentation](https://flutter.dev/).
+
+For instructions integrating Flutter modules to your existing applications,
+see the [add-to-app documentation](https://flutter.dev/docs/development/add-to-app).
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/README.md]: Please update.

File: ostorlab_insecure_flutter_app/lib/bug_rule.dart [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  Future<void> run() async {
+    // Implementation of run method
+  }
+
+  String getDescription();
+
+  String toString() {
+    return runtimeType.toString();
+  }
+}
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/lib/bug_rule.dart]: Please add docstring.

File: ostorlab_insecure_flutter_app/lib/bug_rule.dart [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+abstract class BugRule {
+  late BuildContext context;
+
+  void setContext(BuildContext context) {
+    this.context = context;
+  }
+
+  BuildContext getContext() {
+    return context;
+  }
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/lib/bug_rule.dart]: Dart have a get and set keyword for getters and setters.

File: ostorlab_insecure_flutter_app/test/widget_test.dart [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+// gestures. You can also use WidgetTester to find child widgets in the widget
+// tree, read text, and verify that the values of widget properties are correct.
+
+import 'package:flutter/material.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+import 'package:ostorlab_insecure_flutter_app/main.dart';
+
+void main() {
+  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/test/widget_test.dart]: I don't get the point of this unit test?

**Review by 3asm — CHANGES_REQUESTED** [review]

File: app/build.gradle [hunk 7]
```diff
@@ -37,19 +40,36 @@ android {
...
     implementation 'androidx.appcompat:appcompat:1.4.1'
     implementation 'com.google.android.material:material:1.4.+'
     implementation 'androidx.constraintlayout:constraintlayout:2.1.3'
+    implementation 'androidx.coordinatorlayout:coordinatorlayout:1.2.0'
     androidTestImplementation 'com.android.support.test:runner:1.0.1'
     androidTestImplementation 'com.android.support.test.espresso:espresso-core:3.0.1'
     testImplementation 'junit:junit:4.12'
     testImplementation 'org.mockito:mockito-all:1.+'
     implementation 'androidx.biometric:biometric:1.1.0'
+    debugImplementation 'co.ostorlab.ostorlab_insecure_flutter_app:flutter_debug:1.0'
```
Comment (amine3) [inline app/build.gradle:72]: Where are those dependencies?

**Review by amine3 — CHANGES_REQUESTED** [review]

File: app/build.gradle (hunk 2 above)
Comment (amine3) [reply app/build.gradle:48]: Required to add the flutter dependency

File: app/build.gradle (hunk 7 above)
Comment (amine3) [reply app/build.gradle:72]: Added

File: .gitignore (hunk 1 above)
Comment (BlueSquare1) [reply .gitignore]: There is conflict between `.gitignore` from the android project and the one from the flutter module, I had to force include `ostorlab_insecure_flutter_app` and let it use its own `.gitignore`

Comment (3asm) [reply .gitignore]: In that case can you please a comment to explain it.

File: .idea/libraries/Dart_Packages.xml
Comment (3asm) [inline .idea/libraries/Dart_Packages.xml:1]: .idea need to be excluded.

File: ostorlab_insecure_flutter_app/README.md [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+* TLSTraffic :
+> It was identified that the endpoint supports a combination of cipher suites and Secure Sockets Layer / Transport Layer Security (SSL/TLS) protocols that suffer from known cryptographic weaknesses.
+[link](https://docs.ostorlab.co/kb/TLS_INSECURE_CIPHER/)
+
+* StaticIV:
+> Use of a non-random initialization vector makes the application vulnerable to dictionary attacks.
+[link](https://docs.ostorlab.co/kb/CRYPTO_INSECURE_IV/)
+
+* HardcodedCredsInUrl:
+> [TODO]
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/README.md]: These need to move the main readme.md.

File: ostorlab_insecure_flutter_app/lib/bug_rule.dart [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+  /// Sets the [BuildContext] object associated with this [BugRule] instance.
+  void setContext(BuildContext context) {
+    this.context = context;
+  }
+
+  /// Returns the [BuildContext] object associated with this [BugRule] instance.
+  BuildContext getContext() {
+    return context;
+  }
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/lib/bug_rule.dart]: Please use the setter and getter as temp solutions have a nasty habit of staying forever.

File: ostorlab_insecure_flutter_app/lib/bug_rule_caller.dart [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    return buffer.toString();
+  }
+
+  /// Calls the [run] method of each [BugRule] instance managed by this [BugRuleCaller] instance.
+  Future<void> callRules() async {
+    for (final rule in rules) {
+      try {
+        await rule.run();
+      } catch (e) {
+        print(e.toString());
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/lib/bug_rule_caller.dart]: Need a better formatting to state which rule caused the error.

File: ostorlab_insecure_flutter_app/lib/bugs/command_exec.dart [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+/// A [BugRule] that triggers the execution of arbitrary commands in a Flutter app.
+class CommandExec extends BugRule {
+  /// The tag used to identify instances of this rule.
+  static const String _tag = 'CommandExec';
+
+  /// Trigger the [BugRule]
+  @override
+  Future<void> run() async {
+    String domainName = "google.com";
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/lib/bugs/command_exec.dart]: If it doesn't make sense fix it.

File: ostorlab_insecure_flutter_app/lib/bugs/ecb_cipher_mode.dart [hunk 12]
```diff
@@ -0,0 +1,30 @@
+import 'dart:convert';
+import 'package:encrypt/encrypt.dart';
+import 'package:ostorlab_insecure_flutter_app/bug_rule.dart';
+
+/// A [BugRule] implementation that triggers the use of insecure ECB mode for encryption.
+class ECBCipher extends BugRule {
+  /// The tag used to identify instances of this rule.
+  static const String _tag = "AESCipher";
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/lib/bugs/ecb_cipher_mode.dart]:
```suggestion
  static const String _tag = "ECBCipher";
```

File: ostorlab_insecure_flutter_app/lib/bugs/insecure_commands.dart [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+/// A [BugRule] that triggers the execution of insecure commands from an external storage in a Flutter app.
+class InsecureCommands extends BugRule {
+  /// The tag used to identify instances of this rule.
+  static const String _tag = 'InsecureCommands';
+
+  /// Trigger the [BugRule]
+  @override
+  Future<void> run() async {
+    await executeCommand('chmod 755 test_file', '/data/data/');
+    await executeCommand('ping -c 3 www.ostorlab.co', '/sdcard/ostorlab');
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/lib/bugs/insecure_commands.dart]: Why is this insecure?

File: ostorlab_insecure_flutter_app/pubspec.yaml [hunk 14]
```diff
@@ -0,0 +1,93 @@
+name: ostorlab_insecure_flutter_app
+description: A new Flutter module project.
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/pubspec.yaml]: Needs a better description.

File: ostorlab_insecure_flutter_app/lib/main.dart [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+      _controller.text = _output;
+    });
+
+    BugRuleCaller caller = BugRuleCaller(context);
+    _output += 'Adding rules ...\n';
+    caller.addRule(ECBCipher());
+    caller.addRule(ClearTextTraffic());
+    caller.addRule(TLSTraffic());
+    caller.addRule(StaticIV());
+    caller.addRule(HardcodedCredsInUrl());
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/lib/main.dart:72]: As you just said, this is not up to date.

File: ostorlab_insecure_flutter_app/lib/bugs/webview_insecure_settings.dart [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+  ///
+  /// Creates a new [WebView] with insecure settings, such as enabling JavaScript,
+  /// allowing file access, and allowing access from file URLs.
+  /// Also enables debugging in the WebView and loads the URL "http://www.ostorlab.co".
+  @override
+  Future<void> run() async {
+    WebViewController()
+      ..loadRequest(
+        Uri.parse('http://www.ostorlab.co'),
+      )..setJavaScriptMode(JavaScriptMode.unrestricted);
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/lib/bugs/webview_insecure_settings.dart]: Javascript mode is not a real issue, please check / add debug mode and universal file access.

File: ostorlab_insecure_flutter_app/lib/bugs/static_iv.dart [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+  /// Trigger the bug rule
+  @override
+  Future<void> run() async {
+    String data = "Jan van Eyck was here 1434";
+    String password = "ThisIs128bitSize";
+
+    final plainText = utf8.encode(data);
+    final key = Key.fromUtf8(password);
+    final iv = IV.fromUtf8("0123456789abcdef");
+    final encrypter = Encrypter(AES(key, mode: AESMode.ecb));
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/lib/bugs/static_iv.dart]: If you are doing static IV, please use CBC or the default mode.

File: ostorlab_insecure_flutter_app/lib/bugs/insecure_shared_preferences.dart [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+  Future<void> run() async {
+    String myPreference = "myPreference";
+    SharedPreferences prefs = await SharedPreferences.getInstance();
+    await prefs.setString("PrivateOnly", myPreference);
+    SharedPreferences prefs1 = await SharedPreferences.getInstance();
+    await prefs1.setString("WorldReadableOnly", myPreference);
+    SharedPreferences prefs2 = await SharedPreferences.getInstance();
+    await prefs2.setString("WorldReadableAndWritable", myPreference);
+    SharedPreferences prefs3 = await SharedPreferences.getInstance();
+    await prefs3.setString("WorldReadableAndAppend", myPreference);
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/lib/bugs/insecure_shared_preferences.dart]: How are these world xxx? They simply get added.

File: ostorlab_insecure_flutter_app/lib/bugs/insecure_file_permissions.dart [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+    final documentsDir = '/sdcard/Documents';
+    final file = await File('${documentsDir}/$filename').create();
+    await file.writeAsString('writable all');
+    await _setFilePermissions(file.path, '666');
+  }
+
+  /// Sets the permissions of a file using the chmod command.
+  Future<void> _setFilePermissions(String filePath, String permission) async {
+    await Process.run('chmod', [permission, filePath]);
+  }
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/lib/bugs/insecure_file_permissions.dart]: Either we use the file permissions API on flutter, or the insecure command rule is sufficient. Just double check there are no methods to change the file permissions.

**Review by 3asm — CHANGES_REQUESTED** [review]

File: ostorlab_insecure_flutter_app/lib/bug_rule.dart [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+abstract class BugRule {
+  late BuildContext context;
+
+  /// Runs the bug rule implementation asynchronously.
+  Future<void> run() async {
+    // Implementation of run method
+  }
+
+  /// Returns the description of the bug rule.
+  String getDescription();
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/lib/bug_rule.dart]: This still needs a getter.

File: ostorlab_insecure_flutter_app/lib/bug_rule_caller.dart [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+  late List<BugRule> rules;
+  late BuildContext context;
+
+  /// Creates a new instance of [BugRuleCaller] with the provided [BuildContext].
+  BugRuleCaller(this.context) {
+    rules = [];
+  }
+
+  /// Returns the list of [BugRule] instances managed by this [BugRuleCaller] instance.
+  List<BugRule> getRules() {
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/lib/bug_rule_caller.dart]: Needs a getter.

File: ostorlab_insecure_flutter_app/lib/bugs/insecure_commands.dart (hunk 13 above)
Comment (3asm) [reply ostorlab_insecure_flutter_app/lib/bugs/insecure_commands.dart]: You can remove this.

File: ostorlab_insecure_flutter_app/lib/bugs/sqlite_database_call.dart [hunk 22]
```diff
@@ -0,0 +1,73 @@
+import 'dart:async';
+import 'package:ostorlab_insecure_flutter_app/bug_rule.dart';
+import 'package:path/path.dart';
+import 'package:sqflite/sqflite.dart';
+
+class SQLiteDatabaseCall extends BugRule {
+  @override
+  Future<void> run() async {
```
Comment (3asm) [inline ostorlab_insecure_flutter_app/lib/bugs/sqlite_database_call.dart:9]: This needs an untrusted input.

File: README.markdown [hunk 23]
```diff
@@ -86,8 +82,85 @@ The list of vulnerabilities;
...
+
+* ~~InsecureSharedPreferences :~~
+
+* ~~MobileOnlyDownloadManager :~~
+
+* ~~ParcelableMemoryCorruption :~~
+
+* ~~PackageContextLoadCall :~~
+
+* ~~Implicit PendingIntent Vulnerability :~~
```
Comment (3asm) [inline README.markdown]: What is the status on these, can you please make it clearer in the markdown.

**Review by 3asm — APPROVED** [review]

**Review by 3asm — APPROVED** [review]

**Pull request merged**